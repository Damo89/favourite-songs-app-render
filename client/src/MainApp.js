import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import { FaHome } from 'react-icons/fa';
import axios from 'axios';
import { v4 as uuidv4 } from 'uuid';
import './App.css';

const API_BASE = process.env.REACT_APP_API_BASE_URL;

function MainApp() {
  const [songs, setSongs] = useState([]);
  const [currentIndex, setCurrentIndex] = useState(0);
  const [isBackgroundDark, setIsBackgroundDark] = useState(false);
  const [menuOpen, setMenuOpen] = useState(false);
  const [votes, setVotes] = useState({});
  const [userId, setUserId] = useState(null);
  const [fadeClass, setFadeClass] = useState('fade-in');

  const changeSongWithFade = (newIndex) => {
    if (newIndex < 0 || newIndex >= songs.length) return;

    setFadeClass('fade-out');

    setTimeout(() => {
      setCurrentIndex(newIndex);
      setFadeClass('fade-in');
    }, 400); // must match CSS duration
  };

  useEffect(() => {
    let storedId = localStorage.getItem('userId');
    if (!storedId) {
      storedId = uuidv4();
      localStorage.setItem('userId', storedId);
    }
    setUserId(storedId);
  }, []);

  useEffect(() => {
    axios.get(`${API_BASE}/songs`)
      .then(res => {
        const sorted = [...res.data].sort((a, b) => b.id - a.id);
        setSongs(sorted);
        setCurrentIndex(0);
      })
      .catch(err => console.error('Error fetching songs:', err));
  }, []);

  useEffect(() => {
    const setVh = () => {
      document.documentElement.style.setProperty('--vh', `${window.innerHeight * 0.01}px`);
    };
    setVh();
    window.addEventListener('resize', setVh);
    return () => window.removeEventListener('resize', setVh);
  }, []);

  // Load saved votes from localStorage
  useEffect(() => {
    const storedVotes = JSON.parse(localStorage.getItem('votes')) || {};
    setVotes(storedVotes);
  }, []);

  // Save votes to localStorage whenever they change
  useEffect(() => {
    localStorage.setItem('votes', JSON.stringify(votes));
  }, [votes]);

  const getYouTubeEmbedURL = (url) => {
    const match = url.match(/(?:youtube\.com\/watch\?v=|youtu\.be\/)([a-zA-Z0-9_-]+)/);
    return match ? `https://www.youtube.com/embed/${match[1]}` : null;
  };

  const getSpotifySearchURL = (query) => {
    const encodedQuery = encodeURIComponent(query);
    return `https://open.spotify.com/search/${encodedQuery}`;
  };

  const getImageBrightness = (imageUrl, callback) => {
    const img = new Image();
    img.crossOrigin = 'anonymous';
    img.src = imageUrl;

    img.onload = () => {
      const canvas = document.createElement('canvas');
      canvas.width = canvas.height = 10;
      const ctx = canvas.getContext('2d');
      ctx.drawImage(img, 0, 0, 10, 10);
      const imageData = ctx.getImageData(0, 0, 10, 10).data;

      let colorSum = 0;
      for (let i = 0; i < imageData.length; i += 4) {
        const r = imageData[i];
        const g = imageData[i + 1];
        const b = imageData[i + 2];
        const avg = (r + g + b) / 3;
        colorSum += avg;
      }

      const brightness = colorSum / (10 * 10);
      callback(brightness);
    };
  };

  useEffect(() => {
    if (songs.length > 0 && Object.keys(votes).length === 0) {
      const updatedVotes = {};
      songs.forEach(song => {
        updatedVotes[song.id] = {
          status: 0,
          score: song.votes || 0,
        };
      });
      setVotes(updatedVotes);
    }
  }, [songs, votes]);


  const currentSong = songs[currentIndex];

  useEffect(() => {
    if (currentSong?.album_image_url) {
      getImageBrightness(currentSong.album_image_url, (brightness) => {
        setIsBackgroundDark(brightness < 128);
      });
    }
  }, [currentSong]);

  useEffect(() => {
    if (songs.length > 0) {
      setFadeClass('fade-in');
    }
  }, [songs]);

  useEffect(() => {
    if (fadeClass === 'fade-in') {
      const timeout = setTimeout(() => {
        setFadeClass('');
      }, 800); // matches animation duration

      return () => clearTimeout(timeout);
    }
  }, [fadeClass]);

  const handleNavClick = (index) => {
    if (index === currentIndex) return;

    setFadeClass('fade-out');

    setTimeout(() => {
      setCurrentIndex(index);
      setFadeClass('fade-in');
      setMenuOpen(false);
    }, 400); // matches the CSS transition duration
  };


  // Per-song voting logic
  const voteStatus = votes[currentSong?.id]?.status || 0;
  const voteScore = votes[currentSong?.id]?.score !== undefined
    ? votes[currentSong?.id].score
    : currentSong?.votes || 0;

  const updateVote = async (songId, newStatus) => {
    if (!userId) return;

    const previousStatus = votes[songId]?.status || 0;
    const direction = newStatus === previousStatus ? 0 : newStatus;

    try {
      const res = await axios.post(`http://localhost:5000/songs/${songId}/vote`, {
        direction,
        userId
      });

      if (res.data.votes !== null) {
      setVotes({
        ...votes,
        [songId]: {
          status: direction,
          score: res.data.votes
        },
      });
    }
    } catch (err) {
      console.error('Vote failed:', err);
    }
  };

  const handleUpvote = () => {
    updateVote(currentSong.id, 1);
  };

  const handleDownvote = () => {
    updateVote(currentSong.id, -1);
  };

  return (
    <div>
      <header className="navbar">
        <div className="navbar-container">
          <Link to="/" className="home-icon-link" aria-label="Go to Landing Page">
            <FaHome />
          </Link>

          <div className="logo">My Favourite <span className="highlight">100</span> Songs Ever</div>

          <button className="burger" onClick={() => setMenuOpen(!menuOpen)} aria-label="Toggle menu">
            <span className="bar"></span>
            <span className="bar"></span>
            <span className="bar"></span>
          </button>

          <nav className="nav-links">
            <button onClick={() => handleNavClick(0)}>100–76</button>
            <button onClick={() => handleNavClick(25)}>75–51</button>
            <button onClick={() => handleNavClick(50)}>50–26</button>
            <button onClick={() => handleNavClick(75)}>25–11</button>
            <button onClick={() => handleNavClick(90)}>Top 10</button>
          </nav>

          <nav className={`nav-drawer ${menuOpen ? 'open' : ''}`}>
            <div className="drawer-header">
              My Favourite <span className="highlight">100</span> Songs Ever
            </div>
            <button onClick={() => handleNavClick(0)}>100–76</button>
            <button onClick={() => handleNavClick(25)}>75–51</button>
            <button onClick={() => handleNavClick(50)}>50–26</button>
            <button onClick={() => handleNavClick(75)}>25–11</button>
            <button onClick={() => handleNavClick(90)}>Top 10</button>
          </nav>
        </div>
      </header>

      <div className="container">
        {currentSong && (
          <div className="song-card">
            <div
              className="blurred-bg"
              style={{ backgroundImage: `url(${currentSong.album_image_url})` }}
            ></div>

            <div className="glass-overlay">
              <button
                onClick={() => changeSongWithFade(currentIndex - 1)}
                disabled={currentIndex === 0}
                className="nav-chevron-button nav-left"
                aria-label="Previous"
              >
                <svg xmlns="http://www.w3.org/2000/svg" className="chevron-icon" viewBox="0 0 24 24">
                  <polyline points="15 18 9 12 15 6" />
                </svg>
              </button>

              <button
                onClick={() => changeSongWithFade(currentIndex + 1)}
                disabled={currentIndex === songs.length - 1}
                className="nav-chevron-button nav-right"
                aria-label="Next"
              >
                <svg xmlns="http://www.w3.org/2000/svg" className="chevron-icon" viewBox="0 0 24 24">
                  <polyline points="9 6 15 12 9 18" />
                </svg>
              </button>

              <div className={`song-body ${fadeClass} ${isBackgroundDark ? 'light-text' : 'dark-text'}`}>
                <div className="song-header-group">
                  <div className="song-number">{currentSong.id}</div>
                  <h3 className="song-title">
                    <a
                      href={getSpotifySearchURL(`${currentSong.title} ${currentSong.artist}`)}
                      target="_blank"
                      rel="noopener noreferrer"
                    >
                      {currentSong.title}
                    </a>
                  </h3>
                </div>

                <div className="song-content">
                  <div className="vote-controls-wrapper">
                    <div className="vote-controls">
                      <button
                        className={`vote-button upvote ${voteStatus === 1 ? 'active' : ''}`}
                        onClick={handleUpvote}
                        aria-label="Upvote"
                      >
                        ▲
                      </button>
                      <span className="vote-score">{voteScore}</span>
                      <button
                        className={`vote-button downvote ${voteStatus === -1 ? 'active' : ''}`}
                        onClick={handleDownvote}
                        aria-label="Downvote"
                      >
                        ▼
                      </button>
                    </div>
                  </div>

                  <div className="song-info">
                    <p><strong>Artist:</strong>{' '}
                      <a
                        href={getSpotifySearchURL(currentSong.artist)}
                        target="_blank"
                        rel="noopener noreferrer"
                      >
                        {currentSong.artist}
                      </a>
                    </p>
                    <p><strong>Album:</strong>{' '}
                      <a
                        href={getSpotifySearchURL(`${currentSong.album} ${currentSong.artist}`)}
                        target="_blank"
                        rel="noopener noreferrer"
                      >
                        {currentSong.album}
                      </a>
                    </p>
                    <p><strong>Year:</strong> {currentSong.year_of_release}</p>
                    <p className="song-description">{currentSong.description}</p>
                  </div>

                  {getYouTubeEmbedURL(currentSong.youtube_link) && (
                    <div className="video-wrapper">
                      <iframe
                        loading="lazy"
                        src={getYouTubeEmbedURL(currentSong.youtube_link)}
                        title={currentSong.title}
                        frameBorder="0"
                        allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
                        allowFullScreen
                      ></iframe>
                    </div>
                  )}

                  <div className="bottom-nav-buttons">
                    <button
                      onClick={() => changeSongWithFade(currentIndex - 1)}
                      disabled={currentIndex === 0}
                      className="nav-button"
                    >
                      Previous
                    </button>
                    <button
                      onClick={() => changeSongWithFade(currentIndex + 1)}
                      disabled={currentIndex === songs.length - 1}
                      className="nav-button"
                    >
                      Next
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        )}
      </div>
    </div>
  );
}

export default MainApp;
