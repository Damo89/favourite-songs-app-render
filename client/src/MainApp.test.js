import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import axios from 'axios';
import MainApp from './MainApp';

jest.mock('axios');

const mockSongs = [
  {
    id: 100,
    title: 'Song A',
    artist: 'Artist A',
    album: 'Album A',
    year_of_release: 2010,
    description: 'Description A',
    album_image_url: 'https://via.placeholder.com/300',
    youtube_link: 'https://www.youtube.com/watch?v=test1',
    votes: 10,
  },
  {
    id: 90,
    title: 'Top 10 Song',
    artist: 'Top Artist',
    album: 'Top Album',
    year_of_release: 2022,
    description: 'This is one of my top 10 songs.',
    album_image_url: 'https://via.placeholder.com/300',
    youtube_link: 'https://www.youtube.com/watch?v=top10',
    votes: 99,
  }
];

describe('MainApp', () => {
  beforeEach(() => {
    axios.get.mockResolvedValue({ data: mockSongs });
    localStorage.clear();
  });

  it('renders song data', async () => {
    render(
      <MemoryRouter>
        <MainApp />
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(screen.getByText(/Song A/i)).toBeInTheDocument();
      expect(screen.getByText(/Artist A/i)).toBeInTheDocument();
      expect(screen.getByText(/2010/i)).toBeInTheDocument();
    });
  });

  it('renders song data and handles Top 10 nav button click', async () => {
    render(
      <MemoryRouter>
        <MainApp />
      </MemoryRouter>
    );

    // Wait for the first song to appear
    await waitFor(() => {
      expect(screen.getByText(/Song A/i)).toBeInTheDocument();
      expect(screen.getByText(/Artist A/i)).toBeInTheDocument();
      expect(screen.getByText(/2010/i)).toBeInTheDocument();
    });

    // Click "Top 10" button (just to confirm it's clickable)
    const top10Buttons = screen.getAllByText(/Top 10/i);
    fireEvent.click(top10Buttons[0]);

    // Confirm the original song is still there
    expect(screen.getByText(/Song A/i)).toBeInTheDocument();
  });

  it('renders all section navigation buttons', () => {
    render(<MemoryRouter><MainApp /></MemoryRouter>);
    expect(screen.getAllByText('100â€“76')).toHaveLength(2); // nav bar and drawer
    expect(screen.getAllByText('Top 10')).toHaveLength(2);
  });

  it('displays first song details from mocked data', async () => {
    render(<MemoryRouter><MainApp /></MemoryRouter>);
    await waitFor(() => {
      expect(screen.getByText(/Song A/i)).toBeInTheDocument();
      expect(screen.getByText(/Artist A/i)).toBeInTheDocument();
      expect(screen.getByText(/2010/i)).toBeInTheDocument();
      expect(screen.getByText(/Description A/i)).toBeInTheDocument();
    });
  });

  it('Next and Previous navigation buttons render and respond', async () => {
    render(<MemoryRouter><MainApp /></MemoryRouter>);
    await screen.findByText(/Song A/i); // wait for data

    const nextButtons = screen.getAllByRole('button', { name: /Next/i });
    const prevButtons = screen.getAllByRole('button', { name: /Previous/i });

    fireEvent.click(nextButtons[nextButtons.length - 1]);
    fireEvent.click(prevButtons[prevButtons.length - 1]);

    expect(screen.getByText(/Song A/i)).toBeInTheDocument();
  });

  it('renders YouTube video iframe for song', async () => {
    render(<MemoryRouter><MainApp /></MemoryRouter>);
    const iframe = await screen.findByTitle(/Song A/i);
    expect(iframe).toBeInTheDocument();
  });

  it('renders vote buttons and vote count', async () => {
    render(<MemoryRouter><MainApp /></MemoryRouter>);
    await screen.findByText(/Song A/i);

    expect(screen.getByLabelText(/Upvote/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Downvote/i)).toBeInTheDocument();
    expect(screen.getByText('10')).toBeInTheDocument(); // vote count
  });

  it('matches snapshot', async () => {
    const { asFragment } = render(<MemoryRouter><MainApp /></MemoryRouter>);
    await screen.findByText(/Song A/i);
    expect(asFragment()).toMatchSnapshot();
  });
});
