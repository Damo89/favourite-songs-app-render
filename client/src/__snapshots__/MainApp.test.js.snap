// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainApp matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <header
      class="navbar"
    >
      <div
        class="navbar-container"
      >
        <a
          aria-label="Go to Landing Page"
          class="home-icon-link"
          href="/"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z"
            />
          </svg>
        </a>
        <div
          class="logo"
        >
          My Favourite 
          <span
            class="highlight"
          >
            100
          </span>
           Songs Ever
        </div>
        <button
          aria-label="Toggle menu"
          class="burger"
        >
          <span
            class="bar"
          />
          <span
            class="bar"
          />
          <span
            class="bar"
          />
        </button>
        <nav
          class="nav-links"
        >
          <button>
            100–76
          </button>
          <button>
            75–51
          </button>
          <button>
            50–26
          </button>
          <button>
            25–11
          </button>
          <button>
            Top 10
          </button>
        </nav>
        <nav
          class="nav-drawer "
        >
          <div
            class="drawer-header"
          >
            My Favourite 
            <span
              class="highlight"
            >
              100
            </span>
             Songs Ever
          </div>
          <button>
            100–76
          </button>
          <button>
            75–51
          </button>
          <button>
            50–26
          </button>
          <button>
            25–11
          </button>
          <button>
            Top 10
          </button>
        </nav>
      </div>
    </header>
    <div
      class="container"
    >
      <div
        class="song-card"
      >
        <div
          class="blurred-bg"
          style="background-image: url(https://via.placeholder.com/300);"
        />
        <div
          class="glass-overlay"
        >
          <button
            aria-label="Previous"
            class="nav-chevron-button nav-left"
            disabled=""
          >
            <svg
              class="chevron-icon"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="15 18 9 12 15 6"
              />
            </svg>
          </button>
          <button
            aria-label="Next"
            class="nav-chevron-button nav-right"
          >
            <svg
              class="chevron-icon"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="9 6 15 12 9 18"
              />
            </svg>
          </button>
          <div
            class="song-body fade-in dark-text"
          >
            <div
              class="song-header-group"
            >
              <div
                class="song-number"
              >
                100
              </div>
              <h3
                class="song-title"
              >
                <a
                  href="https://open.spotify.com/search/Song%20A%20Artist%20A"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Song A
                </a>
              </h3>
            </div>
            <div
              class="song-content"
            >
              <div
                class="vote-controls-wrapper"
              >
                <div
                  class="vote-controls"
                >
                  <button
                    aria-label="Upvote"
                    class="vote-button upvote "
                  >
                    ▲
                  </button>
                  <span
                    class="vote-score"
                  >
                    10
                  </span>
                  <button
                    aria-label="Downvote"
                    class="vote-button downvote "
                  >
                    ▼
                  </button>
                </div>
              </div>
              <div
                class="song-info"
              >
                <p>
                  <strong>
                    Artist:
                  </strong>
                   
                  <a
                    href="https://open.spotify.com/search/Artist%20A"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    Artist A
                  </a>
                </p>
                <p>
                  <strong>
                    Album:
                  </strong>
                   
                  <a
                    href="https://open.spotify.com/search/Album%20A%20Artist%20A"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    Album A
                  </a>
                </p>
                <p>
                  <strong>
                    Year:
                  </strong>
                   2010
                </p>
                <p
                  class="song-description"
                >
                  Description A
                </p>
              </div>
              <div
                class="video-wrapper"
              >
                <iframe
                  allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
                  allowfullscreen=""
                  frameborder="0"
                  loading="lazy"
                  src="https://www.youtube.com/embed/test1"
                  title="Song A"
                />
              </div>
              <div
                class="bottom-nav-buttons"
              >
                <button
                  class="nav-button"
                  disabled=""
                >
                  Previous
                </button>
                <button
                  class="nav-button"
                >
                  Next
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
